src/dmd/chkscanf.d

```diff
@@ -0,0 +1,641 @@
+                break;
+
+            case Format.p:      // double pointer
+                if (!(t.ty == Tpointer && tnext.ty == Tpointer))
+                    errorMsg(null, slice, e, "void**", t);
+                break;
+
+            case Format.error:
+                deprecation(loc, "format specifier `\"%.*s\"` is invalid", cast(int)slice.length, slice.ptr);
+                break;
+
+            case Format.percent:
+                assert(0);
+        }
+    }
+    return false;
+}
+
+private:
+
+/* Different kinds of formatting specifications, variations we don't
+   care about are merged. (Like we don't care about the difference between
+   f, e, g, a, etc.)
+ */
+enum Format
+{
+    d,          // pointer to int
+    hhd,        // pointer to signed char
+    hd,         // pointer to short int
+    ld,         // pointer to long int
+    lld,        // pointer to long long int
+    jd,         // pointer to intmax_t
+    zd,         // pointer to size_t
+    td,         // pointer to ptrdiff_t
+    u,          // pointer to unsigned int
+    hhu,        // pointer to unsigned char
+    hu,         // pointer to unsigned short int
+    lu,         // pointer to unsigned long int
+    llu,        // pointer to unsigned long long int
+    ju,         // pointer to uintmax_t
+    g,          // pointer to float
+    lg,         // pointer to double
+    Lg,         // pointer to long double
+    s,          // pointer to char string
+    ls,         // pointer to wchar_t string
+    p,          // double pointer
+    percent,    // %% (i.e. no argument)
+    error,      // invalid format specification
+}
+
+
+/**************************************
+ * Parse the *format specifier* which is of the form:
+ *
+ * `%[*][width][length]specifier`
+ *
+ * Params:
+ *      format = format string
+ *      idx = index of `%` of start of format specifier,
+ *          which gets updated to index past the end of it,
+ *          even if Format.error is returned
+ *      asteriskSubSpec = set if `*` for the sub-specifier
+ *      widthSubSpec = width sub-specifier. Set to `0` if no width
+ * Returns:
+ *      Format
+ */
+pure nothrow @safe
+Format parseFormatSpecifier(scope const char[] format, ref size_t idx,
+        out bool asteriskSubSpec, out string widthSubSpec)
+{
+    auto i = idx;
+    assert(format[i] == '%');
+    const length = format.length;
+    string width;
+
+    Format error()
+    {
+        idx = i;
+        return Format.error;
+    }
+
+    ++i;
+    if (i == length)
+        return error();
+
+    if (format[i] == '%')
+    {
+        idx = i + 1;
+        return Format.percent;
+    }
+
+    // * sub-specifier
+    if (format[i] == '*')
+    {
+        ++i;
+        if (i == length)
+            return error();
+        asteriskSubSpec = true;
+    }
+
+    // width sub-specifier
+    {
+        while (isDigit(format[i]))
+        {
+            width ~= format[i];
```

Comment:
There is no reason to store this as a string, or even store it at all. The only purpose here for the field width is to diagnose an invalid format when used with the `n` specifier.

Reply:
`width` removed, as well as the `out` arg. Should we check for any invalid format with `n`, or is it too strict ?